# Exploring Ebay Car Sales Data

Th aim of this project is to clean the data and analyze the included used car listings. The dataset of used cars is from eBay Kleinanzeigen, a classifieds section of the German eBay website. The aim of this project is to clean the data and analyze the included used car listings.

### Summary of results

After analyzing the data, the key findings are following:

- Amoung the top 20 brands, the brand with the highest mean price is mini, following by audi and mercedes_benz.
- The brand with the lowest mean price is renault. Following by fiat and Opel.
- No obvious relationship between the mean price and the mean mileage of the cars can be found
- If we only consider the mean price and mean mileage, smart, hyundai and skoda are the best options as theyhave relatively low mean mileage but with also low mean price.

For more details, please refer to the the full analysis below.

## Importing Data

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore, it can be found [here](https://data.world/data-society/used-cars-data).

In [2]:
# Read in the data
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
# Quick exploration of the data
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Brief observation from the dataset: 
There are total 20 columns in the dataset, 15 of them are strings and 5 of them are integers. There are total 50000 entries in the dataset. Null values exist in column "vehicleType", "gearbox", "model", "fuelType" and "notRepairedDamage". The explaination of the column names are following:

<font color=red>dateCrawled</font> - When this ad was first crawled. All field-values are taken from this date.
<br>
<font color=red>name</font> - Name of the car.
<br>
<font color=red>seller</font> - Whether the seller is private or a dealer.
<br>
<font color=red>offerType</font> - The type of listing
<br>
<font color=red>price</font> - The price on the ad to sell the car.
<br>
<font color=red>abtest</font> - Whether the listing is included in an A/B test.
<br>
<font color=red>vehicleType</font> - The vehicle Type.
<br>
<font color=red>yearOfRegistration</font> - The year in which the car was first registered.
<br>
<font color=red>gearbox</font> - The transmission type.
<br>
<font color=red>powerPS</font> - The power of the car in PS.
<br>
<font color=red>model</font> - The car model name.
<br>
<font color=red>odometer</font> - How many kilometers the car has driven.
<br>
<font color=red>monthOfRegistration</font> - The month in which the car was first registered.
<br>
<font color=red>fuelType</font> - What type of fuel the car uses.
<br>
<font color=red>brand</font> - The brand of the car.
<br>
<font color=red>notRepairedDamage</font> - If the car has a damage which is not yet repaired.
<br>
<font color=red>dateCreated</font> - The date on which the eBay listing was created.
<br>
<font color=red>nrOfPictures</font> - The number of pictures in the ad.
<br>
<font color=red>postalCode</font> - The postal code for the location of the vehicle.
<br>
<font color=red>lastSeenOnline</font> - When the crawler saw this ad last online.

## Primary observations

In [4]:
# quick obsevation of the data
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Observations:
- All the data in "nrOfPictures" is 0.0, which is a candidate to be dropped.
- almost all of the values in seller and offer_type are the same, which are candidates to be dropped.
- In "price" and "odometer", numeric data are stored as text, e.g. 150,000km, which needs to be cleaned.
- the date_crawled, last_seen, and ad_created columns are all identified as string values.
- The maximun value in "yearOfRegistration" is 9999, which is abnormal, further investigation is needed.
- Missing values exist in column "vehicleType", "gearbox", "model", "fuelType" and "notRepairedDamage". Further investigation is needed.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

### Converting the column names from camelcase to snakecase

In [5]:
# show all columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# update column names
new_col_name = [
    'ad_crawled',
    'name',
    'seller',
    'offer_type',
    'price',
    'abtest',
    'vehicle_type',
    'registration_year',
    'gearbox',
    'powerps',
    'model',
    'odometer', 
    'registration_month', 
    'fuel_type', 
    'brand', 
    'unrepaired_damage', 
    'ad_created', 
    'nr_of_pictures', 
    'postal_code',
    'last_seen'
]

autos.columns = new_col_name
autos.head()

ad_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerps   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The column names were converted from camelcase to snakecase and some of the column names were reworded based on the data dictionary to be more descriptive.

### Converting numeric data which stored as text to numeric data

In the following part, the text data stored in "price" and "odometer" will be converted to numeric data.

In [7]:
#convert price to numeric
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)

#confirm results
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [8]:
#convert odometer to numeric
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

#confirm results
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

### Removing abnormal data

#### Price:

In [9]:
# explore price value
print('Number of unique value in price column: {}'.format(autos['price'].unique().shape))
print()
print('Description:')
print(autos['price'].describe())
print()
print('Rows with highest value:')
print(autos['price'].value_counts().sort_index(ascending=False).head(10))
print()
print('Rows with lowest value:')
print(autos['price'].value_counts().sort_index(ascending=True).head(10))

Number of unique value in price column: (2357,)

Description:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Rows with highest value:
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

Rows with lowest value:
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64


In [10]:
# extract data with price between 100 and 70000
autos = autos[autos["price"].between(100,70000)]

In [11]:
# show after cleaning result
print('Number of unique value in price column: {}'.format(autos['price'].unique().shape))
print()
print('Description:')
print(autos['price'].describe())
print()
print('Rows with highest value:')
print(autos['price'].value_counts().sort_index(ascending=False).head(10))
print()
print('Rows with lowest value:')
print(autos['price'].value_counts().sort_index(ascending=True).head(10))

Number of unique value in price column: (2238,)

Description:
count    48142.000000
mean      5727.668377
std       7167.027211
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7490.000000
max      70000.000000
Name: price, dtype: float64

Rows with highest value:
70000    1
69999    1
69997    1
69993    1
69900    1
69500    1
68900    1
68750    1
68500    1
68300    1
Name: price, dtype: int64

Rows with lowest value:
100    134
110      3
111      2
115      2
117      1
120     39
122      1
125      8
129      1
130     15
Name: price, dtype: int64


Some data in price seem abnormal, so it's decided to remove those data. There are 48142 rows left ranging from 100 to 70000.

#### odometer_km:

In [12]:
# explore odometer
print('Number of unique value in odometer_km column: {}'.format(autos['odometer_km'].unique().shape))
print()
print('Description:')
print(autos['odometer_km'].describe())
print()
print('Rows with highest value:')
print(autos['odometer_km'].value_counts().sort_index(ascending=False).head(10))
print()
print('Rows with lowest value:')
print(autos['odometer_km'].value_counts().sort_index(ascending=True).head(10))

Number of unique value in odometer_km column: (13,)

Description:
count     48142.000000
mean     126054.069212
std       39403.749250
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Rows with highest value:
150000    31207
125000     5032
100000     2097
90000      1732
80000      1410
70000      1214
60000      1146
50000      1003
40000       806
30000       764
Name: odometer_km, dtype: int64

Rows with lowest value:
5000      747
10000     240
20000     744
30000     764
40000     806
50000    1003
60000    1146
70000    1214
80000    1410
90000    1732
Name: odometer_km, dtype: int64


Data in odometer_km seem normal. No removal actions were taken.

### Converting the data into a numerical representation

#### ad_crawled, ad_created, last_seen:

In [13]:
# explore date columns
autos[['ad_crawled','ad_created','last_seen']][0:5]

ad_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

It's clear that the first 10 characters represent the data(e.g. 2016-03-26). To understand the date range, the date values will be extracted, a distribution will be generated, which is sorted by the index.

In [14]:
# extract date
autos['ad_crawled'] = autos['ad_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]

autos[['ad_crawled','ad_created','last_seen']][0:5]

ad_crawled  ad_created   last_seen
0  2016-03-26  2016-03-26  2016-04-06
1  2016-04-04  2016-04-04  2016-04-06
2  2016-03-26  2016-03-26  2016-04-06
3  2016-03-12  2016-03-12  2016-03-15
4  2016-04-01  2016-04-01  2016-04-01

#### Distribution of ad_crawled:

In [15]:
# explore distribution of ad_crawled
autos['ad_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025362
2016-03-06    0.014042
2016-03-07    0.036102
2016-03-08    0.033173
2016-03-09    0.033048
2016-03-10    0.032321
2016-03-11    0.032653
2016-03-12    0.036932
2016-03-13    0.015704
2016-03-14    0.036662
2016-03-15    0.034315
2016-03-16    0.029455
2016-03-17    0.031490
2016-03-18    0.012858
2016-03-19    0.034772
2016-03-20    0.037826
2016-03-21    0.037161
2016-03-22    0.032840
2016-03-23    0.032300
2016-03-24    0.029434
2016-03-25    0.031490
2016-03-26    0.032300
2016-03-27    0.031054
2016-03-28    0.034938
2016-03-29    0.034170
2016-03-30    0.033713
2016-03-31    0.031864
2016-04-01    0.033630
2016-04-02    0.035624
2016-04-03    0.038573
2016-04-04    0.036559
2016-04-05    0.013066
2016-04-06    0.003178
2016-04-07    0.001392
Name: ad_crawled, dtype: float64

The crawled date is distributed from 2016-03-05 to 2016-04-07.

#### Distribution of ad_created:

In [16]:
# explore distribution of ad_created
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000042
2016-02-05    0.000042
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000042
2016-02-14    0.000042
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000042
2016-02-19    0.000062
2016-02-20    0.000042
2016-02-21    0.000062
                ...   
2016-03-09    0.033152
2016-03-10    0.032030
2016-03-11    0.032965
2016-03-12    0.036766
2016-03-13    0.017074
2016-03-14    0.035291
2016-03-15    0.034045
2016-03-16    0.029953
2016-03-17    0.031137
2016-03-18    0.013564
2016-03-19    0.033650
2016-03-20    0.037888
2016-03-21 

The ad creation date is distributed from 2015-06-11 to 2016-04-07. Most of the ad were created after March 2016.

#### Distribution of last_seen:

In [17]:
# explore distribution of last_seen
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004321
2016-03-07    0.005442
2016-03-08    0.007332
2016-03-09    0.009576
2016-03-10    0.010656
2016-03-11    0.012422
2016-03-12    0.023825
2016-03-13    0.008890
2016-03-14    0.012650
2016-03-15    0.015870
2016-03-16    0.016451
2016-03-17    0.028146
2016-03-18    0.007312
2016-03-19    0.015787
2016-03-20    0.020668
2016-03-21    0.020564
2016-03-22    0.021395
2016-03-23    0.018570
2016-03-24    0.019733
2016-03-25    0.019131
2016-03-26    0.016638
2016-03-27    0.015517
2016-03-28    0.020834
2016-03-29    0.022330
2016-03-30    0.024719
2016-03-31    0.023846
2016-04-01    0.022870
2016-04-02    0.024885
2016-04-03    0.025113
2016-04-04    0.024573
2016-04-05    0.124839
2016-04-06    0.221989
2016-04-07    0.132026
Name: last_seen, dtype: float64

The last seen date is distributed from 2016-03-05 to 2016-04-07.

### Abnormal value in registration_year:

In [18]:
# explore registration_year
autos['registration_year'].describe()

count    48142.000000
mean      2004.730921
std         87.970196
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The maximun value of registration_year is 9999 and the minimun value is 1000, which is obviously not normal.

In [19]:
# find out abnormal year
autos[autos["registration_year"]<1900]

ad_crawled                                   name  seller offer_type  \
10556  2016-04-01                             UNFAL_Auto  privat    Angebot   
22316  2016-03-29  VW_Kaefer.__Zwei_zum_Preis_von_einem.  privat    Angebot   
24511  2016-03-17            Trabant__wartburg__Ostalgie  privat    Angebot   
32585  2016-04-02                             UNFAL_Auto  privat    Angebot   
49283  2016-03-15                             Citroen_HY  privat    Angebot   

       price   abtest vehicle_type  registration_year  gearbox  powerps  \
10556    450  control          NaN               1800      NaN     1800   
22316   1500  control          NaN               1000  manuell        0   
24511    490  control          NaN               1111      NaN        0   
32585    450  control          NaN               1800      NaN     1800   
49283   7750  control          NaN               1001      NaN        0   

        model  odometer_km  registration_month fuel_type       brand  \
10556     NaN         5000                   2       NaN  mitsubishi   
22316  kaefer         5000                   0    benzin  volkswagen   
24511     NaN         5000                   0       NaN     trabant   
32585     NaN         5000                   2       NaN  mitsubishi   
49283  andere         5000                   0       NaN     citroen   

      unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
10556              nein  2016-04-01               0        63322  2016-04-01  
22316               NaN  2016-03-29               0        48324  2016-03-31  
24511               NaN  2016-03-17               0        16818  2016-04-07  
32585              nein  2016-04-02               0        63322  2016-04-04  
49283               NaN  2016-03-15               0        66706  2016-04-06

It is safe to remove data with registration year earlier than 1900.

In [20]:
# extract data with registration year later than 1900
autos = autos[autos["registration_year"]>1900]

In [21]:
# find out data with registration year later than 2016
autos[autos["registration_year"]>2016]

ad_crawled                                               name  seller  \
10     2016-03-15                       VW_Golf_Tuning_in_siber/grau  privat   
65     2016-04-04                      Ford_Fiesta_zum_ausschlachten  privat   
68     2016-04-03  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...  privat   
84     2016-03-27                                     Renault_twingo  privat   
113    2016-04-03                              Golf_4_Anfaenger_auto  privat   
164    2016-03-13  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...  privat   
197    2016-04-05                              VW_Polo_9N_an_Bastler  privat   
253    2016-03-27             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017  privat   
348    2016-03-17  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...  privat   
390    2016-03-25   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19  privat   
438    2016-03-10                VW_Golf_5_V_1.6_2004_Klima_Tempomat  privat   
453    2016-03-28                                         Armee_Jeep  privat   
454    2016-03-25    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm  privat   
457    2016-03-15  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...  privat   
477    2016-03-31                                         Citroën_C4  privat   
545    2016-03-22                                      Bmw_x5_/__E53  privat   
548    2016-03-10            Peugeot_206_1.9_Hdi_4_Tueren_Tuev_02.17  privat   
557    2016-03-16                           Smart_Fortwo_Passion_MHD  privat   
568    2016-03-18             PRIMA_ANFÄNGERAUTO_smart_fortwo_cabrio  privat   
577    2016-04-03                                  Volkswagen_Golf_3  privat   
582    2016-03-31                                    V10_TDI_TOUAREG  privat   
654    2016-04-01              Ich_verkaufe_meinen_gepflegten_daewoo  privat   
740    2016-04-02                            Seat_Leon_zu_verkaufen!  privat   
754    2016-03-26                         Ford_Ka_Kleinwagen_BJ_2008  privat   
758    2016-03-12                            Honda_Civic_mit_TÜV_!!!  privat   
765    2016-03-26               BMW_3er_Limousine_mit_1_Jahr_TÜV!!!!  privat   
802    2016-03-19                 Lada_mit_wenig_km_neuem_Tuev_bj_08  privat   
871    2016-04-04                                   Citroen_C2_60_PS  privat   
889    2016-03-12                          Verkaufe_Opel_Corsa_C_1_2  privat   
910    2016-03-15                         Renault_Kangoo_bj.2007_LPG  privat   
...           ...                                                ...     ...   
49108  2016-03-14                  TDI_20_L_DIESEL140_PS..AUTOMATICK  privat   
49130  2016-03-09                                      Renault_Modus  privat   
49153  2016-03-12                                        Corsa_c20xe  privat   
49154  2016-03-24                    Auto_mit_TÜV_bis_September_2017  privat   
49178  2016-03-24       Mazda_323_F_TÜV__neu_2te_Hand_viele_Neuteile  privat   
49185  2016-03-28              Alfa_Romeo_156_Sportwagon_1_8_T_Spark  privat   
49218  2016-04-01          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev  privat   
49245  2016-03-26                         Audi_A3_1.8T_tausch_moegl.  privat   
49259  2016-03-19                  Volkswagen_Lupo_TÜV_06.2017_Klima  privat   
49261  2016-03-10                                        Toyota_aygo  privat   
49262  2016-03-26                                        Opel_vectra  privat   
49266  2016-03-10  Toyota_Corolla_Verso_7_Sitzer_104.000km_2_2_Di...  privat   
49343  2016-03-12                                   bmw_320i_touring  privat   
49347  2016-03-21                         Honda_civic_mit_TÜV_3/2017  privat   
49354  2016-04-05                        Bmw_e39_523i_mit_neuem_Tuev  privat   
49389  2016-03-10                                         Astra_G_CC  privat   
49411  2016-03-28                            Renault_twingo_Tuev_neu  privat   
49522  2016-03-14               Mazda_Demio_zu_verkaufen!_RESERVIERT  privat   
49557  

It's also safe to remove with registration year later than 2016 as a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.

In [22]:
# extract data with registration year earlier than 2016
autos = autos[autos["registration_year"]<=2016]

In [23]:
# confirm result after cleaning
print(autos['registration_year'].describe())
print()
print(autos['registration_year'].value_counts(normalize=True))

count    46270.000000
mean      2002.936568
std          7.107087
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

2000    0.067063
2005    0.062913
1999    0.062222
2004    0.058288
2003    0.058202
2006    0.057640
2001    0.056797
2002    0.053534
1998    0.050486
2007    0.049081
2008    0.047677
2009    0.044889
1997    0.041560
2011    0.034817
2010    0.034191
1996    0.029263
2012    0.028096
2016    0.025848
1995    0.025740
2013    0.017117
2014    0.014199
1994    0.013529
1993    0.009077
2015    0.008040
1992    0.007910
1991    0.007305
1990    0.007154
1989    0.003696
1988    0.002853
1985    0.002010
          ...   
1976    0.000454
1977    0.000454
1969    0.000389
1975    0.000389
1965    0.000367
1964    0.000259
1963    0.000151
1961    0.000130
1959    0.000130
1962    0.000086
1956    0.000086
1937    0.000086
1958    0.000086
1957    0.000043
1954    0.000043
1

After removing date with registration year below 1900 and above 2016, there are 45074 data left. more of the cars are registrated in 2000 (6.88%), following by 2005 (6.46%) and 1999 (6.39%)

## Data analysis

### Exploring variations across different car brands

The average value of the top 20 brands will be explored.

In [24]:
# show top 20 car brands
top_20_brands = autos['brand'].value_counts().head(20).index
top_20_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [25]:
# find out mean price for top 20 brands
top_20_brands_mean_price = {}
for brand in top_20_brands:
    top_20_brands_mean_price[brand] = int(round(autos.loc[autos['brand'] == brand, "price"].mean()))
    
top_20_brands_mean_price

{'audi': 9306,
 'bmw': 8237,
 'citroen': 3796,
 'fiat': 2837,
 'ford': 3719,
 'hyundai': 5411,
 'mazda': 4130,
 'mercedes_benz': 8472,
 'mini': 10639,
 'nissan': 4757,
 'opel': 3005,
 'peugeot': 3114,
 'renault': 2455,
 'seat': 4433,
 'skoda': 6410,
 'smart': 3596,
 'sonstige_autos': 10324,
 'toyota': 5167,
 'volkswagen': 5437,
 'volvo': 4993}

The brand with the highest mean price is mini (USD10654), following by "sonstige_autos", which is not a brand, it means "other cars". The following brands are audi (USD9358) and mercedes_benz (USD8551). The brand with the lowest mean price is renault, the mean price is USD2469 only.

### Comparing average mileage and mean price

Aggregation will be used to understand the average mileage for those cars and if there's any visible link with mean price. 

In [26]:
# assign mean price to new list for comparison
brand_series = pd.Series(top_20_brands_mean_price)
brand_compare = pd.DataFrame(brand_series, columns=['mean_price'])
brand_compare

mean_price
audi                  9306
bmw                   8237
citroen               3796
fiat                  2837
ford                  3719
hyundai               5411
mazda                 4130
mercedes_benz         8472
mini                 10639
nissan                4757
opel                  3005
peugeot               3114
renault               2455
seat                  4433
skoda                 6410
smart                 3596
sonstige_autos       10324
toyota                5167
volkswagen            5437
volvo                 4993

In [27]:
# find out mean mileage 
top_20_brands_mean_mileage = {}
for brand in top_20_brands:
    top_20_brands_mean_mileage[brand] = int(round(autos.loc[autos['brand'] == brand, "odometer_km"].mean()))
    
top_20_brands_mean_mileage

{'audi': 129296,
 'bmw': 132781,
 'citroen': 119555,
 'fiat': 116950,
 'ford': 124320,
 'hyundai': 106886,
 'mazda': 124554,
 'mercedes_benz': 131193,
 'mini': 88309,
 'nissan': 118326,
 'opel': 129384,
 'peugeot': 127128,
 'renault': 128326,
 'seat': 121537,
 'skoda': 110907,
 'smart': 99734,
 'sonstige_autos': 91721,
 'toyota': 115944,
 'volkswagen': 128800,
 'volvo': 138582}

In [28]:
# put mean mileage into comparing list
mile_series = pd.Series(top_20_brands_mean_mileage)
brand_compare['mean_mileage'] = mile_series
brand_compare

mean_price  mean_mileage
audi                  9306        129296
bmw                   8237        132781
citroen               3796        119555
fiat                  2837        116950
ford                  3719        124320
hyundai               5411        106886
mazda                 4130        124554
mercedes_benz         8472        131193
mini                 10639         88309
nissan                4757        118326
opel                  3005        129384
peugeot               3114        127128
renault               2455        128326
seat                  4433        121537
skoda                 6410        110907
smart                 3596         99734
sonstige_autos       10324         91721
toyota                5167        115944
volkswagen            5437        128800
volvo                 4993        138582

We can say that there is no obvious relationship between the mean price and the mean mileage of the cars. mini has the highest mean price and lowest mean mileage, sonstige_autos (other cars) are has the second highest mean price and second lowest mean mileage. Besides this two, smart, hyundai and skoda have relatively low mean mileage but with also low mean price. For other ar brands in high price range (audi, bmw and mercedes benz), they have driven in average more than 129000km. 

## Conclusion

So far we have exlpore and analyze the Ebay Car Sales Data. The key findings are following:
- Amoung the top 20 brands, the brand with the highest mean price is mini (USD10654), following by audi (USD9358) and mercedes_benz (USD8551). 
- The brand with the lowest mean price is renault, the mean price is USD2469 only. Following by fiat (USD2837) and Opel (USD3005)
- No obvious relationship between the mean price and the mean mileage of the cars can be found
- If we only consider the mean price and mean mileage, smart, hyundai and skoda are the best options as theyhave relatively low mean mileage but with also low mean price.